Source/LibationSearchEngine/SearchEngine.cs

```diff
@@ -25,164 +22,47 @@ public class SearchEngine
-
-        // use these common fields in the "all" default search field
-        private static IEnumerable<Func<LibraryBook, string>> allFieldIndexRules { get; }
-            = new List<Func<LibraryBook, string>>
-            {
-                idIndexRules[ALL_AUDIBLE_PRODUCT_ID],
-                stringIndexRules[ALL_TITLE],
-                stringIndexRules[ALL_AUTHOR_NAMES],
-                stringIndexRules[ALL_NARRATOR_NAMES],
-                stringIndexRules[ALL_SERIES_NAMES]
-            };
-        #endregion
 
-        #region get search fields. used for display in help
-        public static IEnumerable<string> GetSearchIdFields()
+        private static bool isAuthorNarrated(Book book)
         {
-            foreach (var key in idIndexRules.Keys)
-                yield return key;
-        }
-
-        public static IEnumerable<string> GetSearchStringFields()
-        {
-            foreach (var key in stringIndexRules.Keys)
-                yield return key;
-        }
-
-        public static IEnumerable<string> GetSearchBoolFields()
-        {
-            foreach (var key in boolIndexRules.Keys)
-                yield return key;
+            var authors = book.Authors.Select(a => a.Name).ToArray();
+            var narrators = book.Narrators.Select(a => a.Name).ToArray();
+            return authors.Intersect(narrators).Any();
         }
 
-        public static IEnumerable<string> GetSearchNumberFields()
+		// use these common fields in the "all" default search field
+		public static IndexRuleCollection FieldIndexRules { get; } = new IndexRuleCollection
         {
-            foreach (var key in numberIndexRules.Keys)
-                yield return key;
-        }
+            { FieldType.ID, Book => Book.AudibleProductId.ToLowerInvariant(), nameof(Book.AudibleProductId), "ProductId", "Id", "ASIN" },
+            { FieldType.Raw, Book => Book.AudibleProductId, _ID_ },
+            { FieldType.String, Book => Book.Title, nameof(Book.Title), "ProductId", "Id", "ASIN" },
+            { FieldType.String, Book => Book.AuthorNames(), "AuthorNames", "Author", "Authors" },
+            { FieldType.String, Book => Book.NarratorNames(), "NarratorNames", "Narrator", "Narrators" },
+            { FieldType.String, Book => Book.Publisher, nameof(Book.Publisher) },
+            { FieldType.String, Book => Book.SeriesNames(), "SeriesNames", "Narrator", "Series" },
+            { FieldType.String, Book => string.Join(", ", Book.SeriesLink.Select(s => s.Series.AudibleSeriesId)), "SeriesId" },
+            { FieldType.String, Book => Book.CategoriesIds() is null ? null : string.Join(", ", Book.CategoriesIds()), nameof(Book.Category), "Categories", "CategoriesId", "CategoryId", "CategoriesNames" },
+            { FieldType.String, Book => Book.UserDefinedItem.Tags, TAGS.FirstCharToUpper() },
+            { FieldType.String, Book => Book.Locale, "Locale", "Region" },
```

Comment:
Ok, I'll bite. How the hell does it know to use the correct `Add` overload here for the Book vs LibraryBook func?

Reply:
Type inference from the methods called on the lambda arguments. I'll admit I was damn impressed too. Roslyn is pretty clever!

Reply:
I've used VS for like 20 years and this is honestly blowing my mind. It just keeps getting better. Hot damn I love coding in C# + VS.

Reply:
> Hot damn I love coding in C# + VS.

+1000, which is what makes development in any other language fell like wading through knee-deep mud.